src/Graph.ts

```diff
@@ -135,6 +135,21 @@ export default class Graph {
 		return ast;
 	}
 
+	ensureModule(module: Module | ExternalModule) {
+		// Make sure that the same module / external module can only be added to the
+		// graph once since it may be requested multiple times over the life of a
+		// service.
+		if (module instanceof Module) {
+			if (this.modules.indexOf(module) === -1) {
+				this.modules.push(module);
+			}
+		} else {
+			if (this.externalModules.indexOf(module) === -1) {
+				this.externalModules.push(module);
+			}
+		}
+	}
```

Comment:
const modules = module instanceof Module ? this.modules : this.externalModules;

if (!modules.includes(module)) {
    modules.push(module);
}

Reply:
what do you think about this way?

Reply:
I like the approach but TypeScript hates it. The result is that `modules` is treated as `Module[] | ExternalModule[]` and therefore the `.includes()` call causes TypeScript to complains with: ```
Argument of type 'Module | ExternalModule' is not assignable to parameter of type 'never'.
  The intersection 'ExternalModule & Module' was reduced to 'never' because property 'options' exists in multiple constituents and is private in some.
    Type 'Module' is not assignable to type 'never'.
```

Reply:
Just tried this and it works:
```ts
		const modules: Array<Module | ExternalModule> = module instanceof Module ? this.modules : this.externalModules;

		if (!modules.includes(module)) {
			modules.push(module);
		}
```

Not attached to either approach.